File: tools/preparar_fotos.py
#!/usr/bin/env python3
"""Prepara una foto para la web: la recorta, la reduce y la guarda en WebP.

    python3 tools/preparar_fotos.py ORIGEN NOMBRE [--relacion 4:3] [--centro 0.5,0.3]
                                                  [--recorte x0,y0,x1,y1]

    ORIGEN   la foto tal como llega (JPG, PNG, la que mande el cliente por WhatsApp)
    NOMBRE   donde queda dentro de public/img, sin extension. Por ejemplo
             fotos/fachada-tejares-del-norte

HERRAMIENTA LOCAL: se corre en el computador de quien edita, no en el servidor.

Por que existe: una foto de celular pesa 300 KB a 1 MB y mide 1500 px o mas.
Subirla tal cual haria lenta la pagina, sobre todo en un celular con datos.
Aqui se recorta a la forma que necesita el sitio, se genera una version chica y
una grande (el navegador elige la que le sirve) y se borran los metadatos.

    --relacion  forma final: 4:3 (tarjetas), 3:2 (fotos anchas), 16:9, 1:1 (circulos)
    --centro    que parte de la foto conservar al recortar, de 0 a 1 en x,y.
                0.5,0.5 es el centro; 0.5,0 es pegado arriba. Por defecto, el centro.
    --recorte   caja exacta en pixeles de la foto original. Manda sobre --relacion.
"""
import argparse
import sys
from pathlib import Path

from PIL import Image, ImageOps

# El sitio muestra tarjetas de ~330 px y fotos anchas de ~760 px, y un celular
# de alta densidad necesita ~1,75 veces eso. El escalon de 900 evita que el
# celular tenga que bajar la version de 1200 cuando le basta con menos.
ANCHOS = (600, 900, 1200)
ANCHO_MINIMO = 600
CALIDAD = 80

# Carpeta de imagenes del sitio, relativa a la raiz del proyecto.
CARPETA_IMG = Path(__file__).resolve().parent.parent / "public" / "img"


def _leer_relacion(texto):
    """'4:3' -> (4, 3). Cualquier otra cosa es un error que se entiende."""
    try:
        ancho, alto = (float(parte) for parte in texto.split(":"))
    except (ValueError, AttributeError):
        raise ValueError(f"La relación '{texto}' no es válida. Se escribe así: 4:3") from None
    if ancho <= 0 or alto <= 0:
        raise ValueError(f"La relación '{texto}' no es válida: ningún lado puede ser cero.")
    return ancho, alto


def caja_de_recorte(ancho, alto, relacion, centro=(0.5, 0.5)):
    """La caja (x0, y0, x1, y1) mas grande que cabe en la foto con esa relacion.

    Conserva todo el ancho o todo el alto, lo que se pueda, y elige que parte
    dejar con `centro`. Nunca se sale de la foto.
    """
    ancho_rel, alto_rel = _leer_relacion(relacion)
    objetivo = ancho_rel / alto_rel

    if ancho / alto > objetivo:              # la foto es mas ancha que lo pedido
        alto_caja, ancho_caja = alto, round(alto * objetivo)
    else:                                    # la foto es mas alta que lo pedido
        ancho_caja, alto_caja = ancho, round(ancho / objetivo)

    cx = min(max(centro[0], 0.0), 1.0)
    cy = min(max(centro[1], 0.0), 1.0)
    x0 = round((ancho - ancho_caja) * cx)
    y0 = round((alto - alto_caja) * cy)
    return x0, y0, x0 + ancho_caja, y0 + alto_caja


def _anchos_de_salida(ancho_recorte):
    """Los tamanos a generar, sin estirar nunca la foto.

    Si el recorte no llega a 1200, el tamano grande es el propio recorte.
    """
    if ancho_recorte < ANCHO_MINIMO:
        raise ValueError(
            f"La foto es muy pequeña ({ancho_recorte} px de ancho tras recortar): "
            f"se vería borrosa. Hace falta al menos {ANCHO_MINIMO} px."
        )
    # Un tamano chico solo vale la pena si es claramente mas chico que el grande;
    # si se parecen, seria el mismo archivo dos veces.
    chicos = {w for w in ANCHOS if w < ancho_recorte * 0.8}
    return sorted(chicos | {min(max(ANCHOS), ancho_recorte)})


def preparar(origen, nombre, destino=CARPETA_IMG, relacion=None, centro=(0.5, 0.5), recorte=None):
    """Recorta y reduce una foto. Devuelve la lista de archivos creados."""
    origen = Path(origen)
    if not origen.exists():
        raise FileNotFoundError(f"La foto '{origen}' no existe.")

    with Image.open(origen) as original:
        # Los celulares guardan la foto "acostada" y anotan como girarla en los
        # metadatos. Si no se corrige aqui, una foto vertical sale horizontal.
        imagen = ImageOps.exif_transpose(original).convert("RGB")

    if recorte:
        caja = tuple(int(v) for v in recorte)
    elif relacion:
        caja = caja_de_recorte(*imagen.size, relacion, centro)
    else:
        caja = (0, 0, *imagen.size)
    imagen = imagen.crop(caja)

    salida = Path(destino) / nombre
    salida.parent.mkdir(parents=True, exist_ok=True)

    archivos = []
    for ancho in _anchos_de_salida(imagen.width):
        alto = round(imagen.height * ancho / imagen.width)
        version = imagen.resize((ancho, alto), Image.LANCZOS) if ancho != imagen.width else imagen
        archivo = salida.with_name(f"{salida.name}-{ancho}.webp")
        # No se pasa exif: el WebP nace sin ubicacion, fecha ni modelo de celular.
        version.save(archivo, "WEBP", quality=CALIDAD, method=6)
        archivos.append(archivo)
    return archivos


def main(argv=None):
    parser = argparse.ArgumentParser(description="Prepara una foto para la web.")
    parser.add_argument("origen")
    parser.add_argument("nombre", help="ruta dentro de public/img, sin extensión")
    parser.add_argument("--relacion", help="forma final, por ejemplo 4:3")
    parser.add_argument("--centro", default="0.5,0.5", help="parte a conservar, por ejemplo 0.5,0.2")
    parser.add_argument("--recorte", help="caja en píxeles: x0,y0,x1,y1")
    args = parser.parse_args(argv)

    try:
        centro = tuple(float(v) for v in args.centro.split(","))
        recorte = tuple(int(v) for v in args.recorte.split(",")) if args.recorte else None
        archivos = preparar(args.origen, args.nombre, relacion=args.relacion,
                            centro=centro, recorte=recorte)
    except (ValueError, FileNotFoundError) as error:
        print(f"No se pudo preparar la foto: {error}")
        return 1

    for archivo in archivos:
        with Image.open(archivo) as im:
            print(f"  {archivo.relative_to(CARPETA_IMG)}  {im.width}x{im.height}  "
                  f"{archivo.stat().st_size // 1024} KB")
    print(f'\nEn el sitio se usa como  "foto": "{args.nombre}"')
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/preparar-fotos.mjs
#!/usr/bin/env node
/**
 * Prepara una foto para la web: la recorta, la reduce y la guarda en WebP.
 *
 *   node tools/preparar-fotos.mjs ORIGEN NOMBRE [--relacion 4:3] [--centro 0.5,0.3]
 *                                              [--recorte x0,y0,x1,y1]
 *
 *   ORIGEN   la foto tal como llega (JPG, PNG, la que mande el cliente por WhatsApp)
 *   NOMBRE   dónde queda dentro de public/img, sin extensión. Por ejemplo
 *            fotos/fachada-tejares-del-norte
 *
 * HERRAMIENTA LOCAL: se corre en el computador de quien edita, nunca en el
 * servidor. Usa Node y sharp, que ya hacen falta para compilar el sitio
 * (Foto.astro los usa): así el cliente no instala nada aparte, en Windows o
 * en Linux/macOS. Reemplaza a la version anterior en Python, que exigia
 * instalar Python y Pillow por separado — un obstaculo que no tenia ningun
 * otro paso del flujo del cliente.
 *
 * Por qué existe: una foto de celular pesa medio mega y mide 1500 px o más.
 * Subirla tal cual haría lenta la página, sobre todo con datos móviles. Aquí
 * se recorta a la forma que necesita el sitio, se genera una versión chica y
 * una grande (el navegador elige la que le sirve) y se borran los metadatos.
 *
 *   --relacion  forma final: 4:3 (tarjetas), 3:2 (fotos anchas), 16:9, 1:1 (círculos)
 *   --centro    qué parte de la foto conservar al recortar, de 0 a 1 en x,y.
 *               0.5,0.5 es el centro; 0.5,0 es pegado arriba. Por defecto, el centro.
 *   --recorte   caja exacta en píxeles de la foto original. Manda sobre --relacion.
 */
import { existsSync } from "node:fs";
import { mkdir, readdir } from "node:fs/promises";
import { dirname, basename, join, relative } from "node:path";
import { fileURLToPath } from "node:url";
import sharp from "sharp";

// El sitio muestra tarjetas de ~330 px y fotos anchas de ~760 px, y un celular
// de alta densidad necesita ~1,75 veces eso. El escalón de 900 evita que el
// celular tenga que bajar la versión de 1200 cuando le basta con menos.
const ANCHOS = [600, 900, 1200];
const ANCHO_MINIMO = 600;
const CALIDAD = 80;

// Carpeta de imágenes del sitio, relativa a la raíz del proyecto.
const RAIZ = dirname(dirname(fileURLToPath(import.meta.url)));
const CARPETA_IMG = join(RAIZ, "public", "img");

function leerRelacion(texto) {
  const partes = String(texto ?? "").split(":").map(Number);
  const [ancho, alto] = partes;
  if (partes.length !== 2 || !Number.isFinite(ancho) || !Number.isFinite(alto) || ancho <= 0 || alto <= 0) {
    throw new Error(`La relación '${texto}' no es válida. Se escribe así: 4:3`);
  }
  return ancho / alto;
}

/**
 * La caja {x0, y0, x1, y1} más grande que cabe en la foto con esa relación.
 * Conserva todo el ancho o todo el alto, lo que se pueda, y elige qué parte
 * dejar con `centro`. Nunca se sale de la foto.
 */
export function cajaDeRecorte(ancho, alto, relacion, centro = [0.5, 0.5]) {
  const objetivo = leerRelacion(relacion);

  let anchoCaja, altoCaja;
  if (ancho / alto > objetivo) {
    altoCaja = alto;
    anchoCaja = Math.round(alto * objetivo);
  } else {
    anchoCaja = ancho;
    altoCaja = Math.round(ancho / objetivo);
  }

  const cx = Math.min(Math.max(centro[0], 0), 1);
  const cy = Math.min(Math.max(centro[1], 0), 1);
  const x0 = Math.round((ancho - anchoCaja) * cx);
  const y0 = Math.round((alto - altoCaja) * cy);
  return { x0, y0, x1: x0 + anchoCaja, y1: y0 + altoCaja };
}

/** Los tamaños a generar, sin estirar nunca la foto. */
function anchosDeSalida(anchoRecorte) {
  if (anchoRecorte < ANCHO_MINIMO) {
    throw new Error(
      `La foto es muy pequeña (${anchoRecorte} px de ancho tras recortar): se vería ` +
        `borrosa. Hace falta al menos ${ANCHO_MINIMO} px.`
    );
  }
  // Un tamaño chico solo vale la pena si es claramente más chico que el
  // grande; si se parecen, sería el mismo archivo dos veces.
  const chicos = ANCHOS.filter((w) => w < anchoRecorte * 0.8);
  return [...new Set([...chicos, Math.min(Math.max(...ANCHOS), anchoRecorte)])].sort((a, b) => a - b);
}

/** Recorta y reduce una foto. Devuelve la lista de rutas creadas. */
export async function preparar(origen, nombre, destino = CARPETA_IMG, opciones = {}) {
  const { relacion, centro = [0.5, 0.5], recorte } = opciones;

  if (!existsSync(origen)) {
    throw new Error(`La foto '${origen}' no existe.`);
  }

  // .rotate() sin argumentos lee la orientación EXIF del celular y gira la
  // imagen para que quede derecha; sharp no lo hace solo. El resto de la
  // cadena no vuelve a leer el EXIF, así que la salida nace sin metadatos:
  // no hay que borrarlos aparte.
  let imagen = sharp(origen).rotate();
  const metadatos = await imagen.metadata();
  let ancho = metadatos.width;
  let alto = metadatos.height;
  // Si venía girada 90°/270°, sharp ya intercambió ancho y alto al rotar,
  // pero metadata() sigue reportando los originales: hay que corregirlo.
  if ([5, 6, 7, 8].includes(metadatos.orientation)) [ancho, alto] = [alto, ancho];

  let caja;
  if (recorte) {
    const [x0, y0, x1, y1] = recorte;
    caja = { x0, y0, x1, y1 };
  } else if (relacion) {
    caja = cajaDeRecorte(ancho, alto, relacion, centro);
  } else {
    caja = { x0: 0, y0: 0, x1: ancho, y2: alto, y1: alto };
  }
  imagen = imagen.extract({
    left: caja.x0,
    top: caja.y0,
    width: caja.x1 - caja.x0,
    height: caja.y1 - caja.y0,
  });
  const anchoRecorte = caja.x1 - caja.x0;
  const altoRecorte = caja.y1 - caja.y0;

  const carpetaSalida = join(destino, dirname(nombre));
  await mkdir(carpetaSalida, { recursive: true });
  const base = basename(nombre);

  const archivos = [];
  for (const anchoSalida of anchosDeSalida(anchoRecorte)) {
    const altoSalida = Math.round((altoRecorte * anchoSalida) / anchoRecorte);
    const archivo = join(carpetaSalida, `${base}-${anchoSalida}.webp`);
    const buffer = await imagen.clone().toBuffer(); // .toBuffer() no muta `imagen`, así se reusa el recorte
    await sharp(buffer)
      .resize({ width: anchoSalida, height: altoSalida, fit: "fill" })
      .webp({ quality: CALIDAD, effort: 6 })
      .toFile(archivo);
    archivos.push(archivo);
  }
  return archivos;
}

async function main(argv) {
  const [origen, nombre, ...resto] = argv;
  if (!origen || !nombre) {
    console.log(
      "Uso: node tools/preparar-fotos.mjs ORIGEN NOMBRE [--relacion 4:3] " +
        "[--centro 0.5,0.5] [--recorte x0,y0,x1,y1]"
    );
    return 1;
  }

  const args = Object.fromEntries(
    resto.reduce((pares, valor, i, arr) => {
      if (valor.startsWith("--")) pares.push([valor.slice(2), arr[i + 1]]);
      return pares;
    }, [])
  );
  const centro = args.centro ? args.centro.split(",").map(Number) : [0.5, 0.5];
  const recorte = args.recorte ? args.recorte.split(",").map(Number) : undefined;

  try {
    const archivos = await preparar(origen, nombre, CARPETA_IMG, {
      relacion: args.relacion,
      centro,
      recorte,
    });
    for (const archivo of archivos) {
      const meta = await sharp(archivo).metadata();
      const kb = Math.round((await sharp(archivo).toBuffer()).length / 1024);
      console.log(`  ${relative(CARPETA_IMG, archivo)}  ${meta.width}x${meta.height}  ${kb} KB`);
    }
    console.log(`\nEn el sitio se usa como  "foto": "${nombre}"`);
    return 0;
  } catch (error) {
    console.log(`No se pudo preparar la foto: ${error.message}`);
    return 1;
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main(process.argv.slice(2)).then((codigo) => process.exit(codigo));
}
